bootstrap/api/v1alpha1/rke2config_types.go

```diff
@@ -17,48 +17,593 @@ limitations under the License.
 package v1alpha1
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 )
 
 // EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
-// Rke2ConfigSpec defines the desired state of Rke2Config
-type Rke2ConfigSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
+// RKE2ConfigSpec defines the desired state of RKE2Config.
+type RKE2ConfigSpec struct {
+	// Files specifies extra files to be passed to user_data upon creation.
+	//+optional
+	Files []File `json:"files,omitempty"`
 
-	// Foo is an example field of Rke2Config. Edit rke2config_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// PreRKE2Commands specifies extra commands to run before rke2 setup runs.
+	//+optional
+	PreRKE2Commands []string `json:"preRKE2Commands,omitempty"`
+
+	// PostRKE2Commands specifies extra commands to run after rke2 setup runs.
+	//+optional
+	PostRKE2Commands []string `json:"postRKE2Commands,omitempty"`
+
+	// AgentConfig specifies configuration for the agent nodes.
+	//+optional
+	AgentConfig RKE2AgentConfig `json:"agentConfig,omitempty"`
+
+	// ServerConfig specifies configuration for the agent nodes.
+	//+optional
+	ServerConfig RKE2ServerConfig `json:"serverConfig,omitempty"`
+
+	// PrivateRegistriesConfig defines the containerd configuration for private registries and local registry mirrors.
+	//+optional
+	PrivateRegistriesConfig Registry `json:"privateRegistriesConfig,omitempty"`
+
+	// Version specifies the rke2 version.
+	//+optional
+	Version string `json:"version,omitempty"`
```

Comment:
What would we use this version for? Would it be used to download rke2 if its not on the base image?

Reply:
Yes, the idea was to do that. RKE2 accepts an environment variable when running its installation script that specifies the version.

It might be possible to get the version from the the `Cluster` object for example, but the exact RKE2 version is slightly different than Kubernetes. If Kubernetes has for example a version  `v1.25.5`, RKE2 might have a version name such as `v1.25.5+rke2r1` or `v1.25.5+rke2r2`.

Reply:
Reading it from `Cluster` should be fine, even if RKE2 versions are different.

Reply:
If its for the Kubernetes version then its fine to keep it here, as long as we are importing this inlining/using this struct in the control plane api definition

Reply:
If we have 2 versions of RKE2 for a Kubernetes version, how do we know which exact version the user wanted?
We might for example query GitHub for the latest release of RKE2 for that particular Kubernetes version (let's say v1.25.5+rke2r2), but what if the user wants v1.25.5+rke2r1 because that's what's on the Rancher Support Matrix ?

So, I would be of the opinion that it is more a specific RKE2 version, not only a Kubernetes version. It might however need validation on the Controller level.

Reply:
I would assume that the version field for our provider always means an RKE2 version.

Reply:
Interesting, i would not make that assumption myself.